### ISSUE TITLE
NLL erroneous Warning: variable does not need to be mutable

### ISSUE THREAD

**Issue by rukai** [rust-lang/rust#50461]
```rust
#![feature(nll)]
struct Foo {
    pub value: i32
}

// triggers bug
fn use_foo_mut(mut foo: Foo) {
    foo = foo;
    println!("{}", foo.value);
}

// does not trigger bug
fn use_i32_mut(mut i: i32) {
    i = i;
    println!("{}", i);
}

fn main() {
    use_foo_mut(Foo { value: 413 });
    use_i32_mut(413);
}
```

gives:
```
warning: variable does not need to be mutable
 --> src/main.rs:8:16
  |
8 | fn use_foo_mut(mut foo: Foo) {
  |                ----^^^
  |                |
  |                help: remove this `mut`
  |
  = note: #[warn(unused_mut)] on by default
```

If I remove the mut then I get a compile error.

Compiled with nightly-2018-05-05

This is not related to https://github.com/rust-lang/rust/issues/50343 as that has been fixed in the nightly I am using.

**Comment by pnkfelix** [comment]
this is probably related to the recent changes to do the use-of-mut checks without requiring AST-borrowck.

**Comment by pnkfelix** [comment]
i.e. this is probably related to PR #48605

**Comment by KiChjang** [comment]
cc me

**Comment by KiChjang** [comment]
Hmm... I think this bug is highlighted by @pnkfelix a couple of days ago, and it's regarding this branch:

https://github.com/rust-lang/rust/blob/8ff4b42064b374bb62043f7729f84b6d979c7667/src/librustc_mir/borrow_check/mod.rs#L1711-L1716

Or perhaps in the branch further down as well:

https://github.com/rust-lang/rust/blob/8ff4b42064b374bb62043f7729f84b6d979c7667/src/librustc_mir/borrow_check/mod.rs#L1805-L1810

I suspect it's the latter, because the MIR generated by the code above contains two move assignment statements. Either way, these two branches should really contain the used mut variables logic.
